OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/alias/AllocationSiteAndTacBasedAliasAnalysis.scala

```diff
@@ -0,0 +1,68 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj.tac.fpcf.analyses.alias
+
+import org.opalj.br.PC
+import org.opalj.br.fpcf.properties.Context
+import org.opalj.br.fpcf.properties.NoContext
+import org.opalj.tac.Goto
+
+trait AllocationSiteAndTacBasedAliasAnalysis extends AllocationSiteBasedAliasAnalysis with TacBasedAliasAnalysis {
+
+    override protected[this] type AnalysisState <: AllocationSiteBasedAliasAnalysisState with TacBasedAliasAnalysisState
+
+    override protected[this] def checkMustAlias(intersection: AllocationSiteBasedAliasSet)(implicit
+        state:   AnalysisState,
+        context: AnalysisContext
+    ): Boolean = {
+
+        val pointsTo1 = state.pointsTo1
+        val pointsTo2 = state.pointsTo2
+
+        if (intersection.size != 1 ||
+            pointsTo1.size != 1 ||
+            pointsTo2.size != 1
+        ) return false
+
+        // they refer to the same allocation site but aren't necessarily the same object (e.g. if the allocation site
+        // is inside a loop or a different method and is executed multiple times)
+
+        val (pointsToContext: Context, pc: PC) = pointsTo1.allPointsTo.head
+        val method = pointsToContext match {
+            case NoContext => context.element1.declaredMethod
+            case _         => pointsToContext.method
+        }
+
+        if (context.element1.isAliasUVar &&
+            context.element2.isAliasUVar &&
+            context.element1.declaredMethod == method &&
+            context.element2.declaredMethod == method
+        ) {
+
+            // Both elements are uVars that point to the same allocation site and both are inside the method of the allocation site
+            // -> they must alias if the allocation site is executed only once (i.e. no loop or recursion)
+
+            val defSite1 = context.element1.asAliasUVar.persistentUVar.defSites
+            val defSite2 = context.element2.asAliasUVar.persistentUVar.defSites
+
+            if (defSite1.size != 1 || defSite1.size != 1 || defSite1.head != defSite2.head) return false // multiple or different def sites for one element -> might be different objects (e.g. due to recursion via parameter)
+
+            val tac = state.tacai1.get
+
+            // the definition site is not the allocation site -> it is a method call or something similar
+            if (pc != defSite1.head) return false
+
+            for (stmt <- state.tacai1.get.stmts) {
+                stmt match {
+                    case goto: Goto =>
+                        val targetPC = tac.stmts(goto.targetStmt).pc
+                        if (targetPC <= pc && goto.pc >= pc) return false // jumping from behind the allocation site in front of it -> might break aliasing because allocation site is executed multiple times
```

Comment:
This is inefficient and incorrect:

Inefficient:
 - Checking every statement whether it is a goto, instead of using already available information (CFG, dominator tree)
 - Matching on the type of the instruction instead of its astID, which allows for an integer check instead of a type check
 - Extracting pcs instead of just working with stmt ids, which are ordered the same way

Incorrect:
 - Would interpret some linear control flow as loops (not unsound, but still):
```
...
5: goto 8
6: allocation site
7: goto 9
8: goto 6
...
```
 - Misses other branching instructions, such as if and switch, so it is unsound as it may miss loops

A better way to do this is to check whether a jump is dominated by its target. That is still not perfect (it misses non-natural loops). OPAL provides facilities to compute dominator trees and check dominance.

Reply:
If I understand it correct, a better way to do it would be iterate over all basic block with a endPC >= allocationSitePC (don't care what previous bb's do) and check whether they are dominated by a bb in front of the allocation site? I don't see a way to directly check "whether a jump is dominated by its target". That would still require me to iterate over all goto statements to identify them. Or should i keep iterating over the statements and only perform the check, whether the goto is problematic, using the dominatorTree? Regarding the implementation: To check the dominance I only found the DominatorTree.foreachDom(int)(int => U) method. How can I map the given int back to the associated bb to check whether that is a problematic dominance?

Reply:
You need to do basic blocks before the allocation, too. See the example above how (linear control flow) jumps can reorder code irrespective of program counters.
Jumps are only at the end of basic blocks, so you don't need to look for gotos (and also, you get all kinds of jumps not just gotos).
There is method `strictlyDominates` on DominatorTree (strictly just means it excludes the BB itself).
I can't come up with the correct procedure from my head just now, but I suppose it will be something like: Is there a dominator (loop header) of the allocation (loop body) which has a predecessor that it dominates (back edge).

Reply:
Sounds good. So far I have this code:

```
// check if the allocation site is dominated by a loop header, i.e., is executed multiple times
domTree.foreachDom(cfg.bb(tac.properStmtIndexForPC(pc)).nodeId)(dom => {

    // check if the dominator is a loop header
    cfg.foreachPredecessor(cfg.allBBs.toSeq(dom).nodeId)(pred => {

        // if the dominator itself dominates one of its predecessors, it is a loop header
        if (domTree.strictlyDominates(dom, cfg.bb(pred).nodeId)) {
            return false
        }
    })

}) 
```
It checks whether the allocation site is dominated by a loop header. It works so far, but overapproximates cases where the allocation site is after the loop and not inside of it. E.g.:
```
public static void mustAliasLoopInFront() {
    for (int i = 0; i < 10; i++) {
        [...]
    }

    Object o2 = new Object();

    o2.hashCode();
}
```
It computes that (o2,o2) is a may alias because the definition site is dominated by a loop header. Do you know whether it is possible to detect such cases, e.g., by checking if the allocation site is behind all predecessors of the loop header? Or would that cause other problems?

Reply:
I guess also check that the backedge is a postdominator of the allocation site?

Reply:
I have tried to access the postDominator tree by adding the following method to the cfg class:

```
def postDominatorTree: PostDominatorTree = {
    val exitNodes = Seq.empty[Int] ++
        (if (normalReturnNode.predecessors.nonEmpty) Seq(normalReturnNode.nodeId) else Seq.empty) ++
        (if (abnormalReturnNode.predecessors.nonEmpty) Seq(abnormalReturnNode.nodeId) else Seq.empty)

    PostDominatorTree(
        if (exitNodes.size == 1) exitNodes.headOption else None,
        exitNodes.contains,
        IntTrieSet.empty,
        exitNodes.foreach,
        foreachSuccessor,
        foreachPredecessor,
        basicBlocks.last.endPC
    )
}
```
However, I get an exception when I try to use it:

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 10
	at org.opalj.graphs.DominatorTree$.$anonfun$create$1(DominatorTree.scala:192)
	at org.opalj.br.cfg.CFG.$anonfun$postDominatorTree$2(CFG.scala:575)
	at org.opalj.br.cfg.CFG.$anonfun$postDominatorTree$2$adapted(CFG.scala:575)
	at org.opalj.graphs.PostDominatorTree$.$anonfun$apply$3(PostDominatorTree.scala:155)
	at org.opalj.graphs.PostDominatorTree$.$anonfun$apply$3$adapted(PostDominatorTree.scala:155)
	at org.opalj.graphs.DominatorTree$.create(DominatorTree.scala:191)
	at org.opalj.graphs.PostDominatorTree$.apply(PostDominatorTree.scala:183)
	at org.opalj.br.cfg.CFG.postDominatorTree(CFG.scala:578)
        [...]
```

Do you know another way to access the PostDominatorTree or what the problem with this one could be?

Reply:
Not sure where that foreachSuccessor method you're using comes from, but apparently it returns -1 when it shouldn't. You can have a look at method postDominatorTree in org.opalj.ai.domain.RecordCFG to check how to properly compute post-dominator trees.

Reply:
I have now found a way to access the postDominatorTree. It currently uses the same way used in the TACAIProvider class. But i am not sure if it is possible to reuse the results used for the tacai instead of calculating them again.

Reply:
I think that is okay for now. We can check whether that computation is performance-sensitive for the analysis, but I don't expect it to be as long as you don't recompute it multiple times per method.

Reply:
I have just implemented that they are being cached after the first computation

Reply:
Note that TACode.cfg.dominatorTree is also recomputed every time and that doesn't seem to be a problem usually.

I think the code you have now will produce different kinds of trees for the dominator and post dominators: the dominator tree from the CFG should be a TAC-based one with Stmts in it, while the post dominator tree you compute is bytecode-based with bytecode Instructions in it, right? That might cause issues.

Reply:
Good point, that's true (interesting that it still passes all tests* :D). I can of course map the values back to their pcs for the postdominator tree like so: `postDomTree.strictlyDominates(cfg.bb(dom).startPC, cfg.bb(allocBB).startPC)`. But a version of the PostDominatorTree that is based on the tac would probably be better.

* They do fail if I change something else